File: ui/src/api/persistent_volume_claim.js
import request from '@/utils/request'

export function listPersistentVolumeClaim(cluster) {
  return request({
    url: `pvc/${cluster}`,
    method: 'get',
  })
}

export function getPersistentVolumeClaim(cluster, namespace, name, output='') {
  return request({
    url: `pvc/${cluster}/${namespace}/${name}`,
    method: 'get',
    params: { output }
  })
}

export function updatePersistentVolumeClaim(cluster, namespace, name, yaml) {
  return request({
    url: `pvc/${cluster}/update/${namespace}/${name}`,
    method: 'post',
    data: { yaml }
  })
}

export function deletePersistentVolumeClaims(cluster, data) {
  return request({
    url: `pvc/${cluster}/delete`,
    method: 'post',
    data: data
  })
}

export function buildPvc(pvc) {
  let size = ''
  console.log(pvc)
  if(pvc.spec.resources && pvc.spec.resources.requests && pvc.spec.resources.requests.storage) {
    size = pvc.spec.resources.requests.storage
  }
  var bp = {
    uid: pvc.metadata.uid,
		name:         pvc.metadata.name,
		status:       pvc.status.phase,
		access_modes:  pvc.spec.accessModes,
		create_time:   pvc.metadata.creationTimestamp,
		namespace:    pvc.metadata.namespace,
		storage_class: pvc.spec.storageClassName ? pvc.spec.storageClassName : '',
		capacity:     size,
  }
  return bp
}

File: pkg/kubernetes/resource/persistent_volume_claim.go
package resource

import (
	"github.com/kubespace/kubespace/pkg/kubernetes/config"
	"github.com/kubespace/kubespace/pkg/kubernetes/types"
	corev1 "k8s.io/api/core/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
	"k8s.io/apimachinery/pkg/runtime"
	"k8s.io/apimachinery/pkg/runtime/schema"
)

var PersistentVolumeClaimGVR = &schema.GroupVersionResource{
	Group:    "",
	Version:  "v1",
	Resource: "persistentvolumeclaims",
}

type PersistentVolumeClaim struct {
	*Resource
}

func NewPersistentVolumeClaim(config *config.KubeConfig) *PersistentVolumeClaim {
	p := &PersistentVolumeClaim{}
	p.Resource = NewResource(config, types.PersistentVolumeClaimType, PersistentVolumeClaimGVR, p.listObjectProcess)
	p.actions = map[string]ActionHandle{
		types.ListAction:   p.List,
		types.GetAction:    p.Get,
		types.DeleteAction: p.Delete,
		types.UpdateAction: p.Update,
	}
	return p
}

type BuildPersistentVolumeClaim struct {
	UID           string                               `json:"uid"`
	Name          string                               `json:"name"`
	Namespace     string                               `json:"namespace"`
	Status        string                               `json:"status"`
	StorageClass  *string                              `json:"storage_class"`
	Capacity      string                               `json:"capacity"`
	AccessModes   []corev1.PersistentVolumeAccessMode  `json:"access_modes"`
	CreateTime    metav1.Time                          `json:"create_time"`
	ReclaimPolicy corev1.PersistentVolumeReclaimPolicy `json:"reclaim_policy"`
}

func (p *PersistentVolumeClaim) ToBuildPersistentVolumeClaim(pvc *corev1.PersistentVolumeClaim) *BuildPersistentVolumeClaim {
	if pvc == nil {
		return nil
	}
	var volumeSize string
	if size, ok := pvc.Spec.Resources.Requests["storage"]; !ok {
		volumeSize = ""
	} else {
		volumeSize = size.String()
	}

	storageClass := pvc.Spec.StorageClassName

	pvcData := &BuildPersistentVolumeClaim{
		UID:          string(pvc.UID),
		Name:         pvc.Name,
		Status:       string(pvc.Status.Phase),
		AccessModes:  pvc.Spec.AccessModes,
		CreateTime:   pvc.CreationTimestamp,
		Namespace:    pvc.Namespace,
		StorageClass: storageClass,
		Capacity:     volumeSize,
	}

	return pvcData
}

func (p *PersistentVolumeClaim) listObjectProcess(query *QueryParams, obj *unstructured.Unstructured) (interface{}, error) {
	pvc := &corev1.PersistentVolumeClaim{}
	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(obj.Object, pvc); err != nil {
		return nil, err
	}
	return p.ToBuildPersistentVolumeClaim(pvc), nil
}
